src/Telemetry/AzureDevOpsClient/AzureDevOpsClient.cs

```diff
@@ -155,36 +163,83 @@ public async Task<WorkItem> CreateRcaWorkItem(string project, string title, Canc
         return JsonConvert.DeserializeObject<WorkItem>(json);
     }
 
-    public async Task<string> TryGetImageName(
+    /// <summary>
+    /// The method reads the logs as a stream, line by line and tries to match the regexes in order, one regex per line. 
+    /// If the consecutive regexes match the lines, the last match is returned.
+    /// </summary>
+    public async Task<string?> GetImageName(
         string logUri,
-        Func<string, string> findImageName,
+        List<Regex> regexes,
         CancellationToken cancellationToken)
     {
-
         using var request = new HttpRequestMessage(HttpMethod.Get, logUri);
         request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("text/plain"));
 
         using var response = await _httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
         using Stream logStream = await response.Content.ReadAsStreamAsync(cancellationToken);
         using StreamReader reader = new StreamReader(logStream);
-        string line;
-        while ((line = reader.ReadLine()) != null)
+        string? line;
+        Queue<string> lineCache = new(regexes.Count);
+        // need to check if a new line will be loaded when the cache is full, it shouldn't because the first condition will tell it to exit the loop
+        while ((line = await reader.ReadLineAsync()) != null && lineCache.Count < regexes.Count())
+        {
+            lineCache.Enqueue(line);
+        }
+
+        // Check if we didn't even have enough lines to fill the cache
+        if (lineCache.Count < regexes.Count())
         {
-            var imageName = findImageName(line);
-            if (imageName != null)
+            return null;
+        }
+
+        string? result;
+        do
+        {
+            result = CheckLineCache(lineCache, regexes);
+
+            if (result != null)
             {
-                return imageName;
+                return result;
+            }
+
+            lineCache.Dequeue();
+            if (line != null)
+            {
+                lineCache.Enqueue(line);
             }
         }
+        while ((line = await reader.ReadLineAsync()) != null);
+
+        // This will return the value if it finds something in the last cache, or null if not
+        return CheckLineCache(lineCache, regexes);
+    }
+
+    private string? CheckLineCache(IEnumerable<string> lineCache, IEnumerable<Regex> regexes)
+    {
+        string? result = null;
 
-        return null;
+        return lineCache.Zip(regexes, (line, regex) => (line, regex))
```

Comment:
Do we need the lambda?

https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.zip?view=net-7.0#system-linq-enumerable-zip-2(system-collections-generic-ienumerable((-0))-system-collections-generic-ienumerable((-1)))

This one hints you could just do `lineCache.Zip(regexes).All(...`

Reply:
Right, gotcha